refactor(docs): enhance the docs sidebar UI

Devanshusharma2005 opened the pull request:
## I have read the [CONTRIBUTING.md](https://github.com/supabase/supabase/blob/master/CONTRIBUTING.md) file.

YES/NO

## What kind of change does this PR introduce?

fixes : #38861

Before : 
<img width="1780" height="1115" alt="Screenshot 2025-09-19 at 9 07 54 PM" src="https://github.com/user-attachments/assets/8b7dc74f-64b0-472d-87e7-1eb5b4a5fd7d" />


After : 


https://github.com/user-attachments/assets/80f502e8-80e5-4a52-bf84-28a3025c77ea



## What is the current behavior?

Please link any relevant issues here.

## What is the new behavior?

Feel free to include screenshots if it includes visual changes.

## Additional context

Add any other context or screenshots.

Devanshusharma2005 committed:
refactor(docs): enhance NavigationMenuGuideListItems component

Devanshusharma2005 committed:
cleanup..

Devanshusharma2005 committed:
Resolve formatting conflicts in NavigationMenuGuideListItems

Devanshusharma2005 committed:
Came up with a crazy recursive approach...

Devanshusharma2005 committed:
some changes.

Devanshusharma2005 committed:
More cleanups...

Devanshusharma2005 committed:
more cleanups....

Devanshusharma2005 committed:
Merge branch 'master' into docs/fix/sidebar-dropdown-indicator

Devanshusharma2005 committed:
formatting the stuff

Devanshusharma2005 committed:
refactor(navigation): remove sample menu data and optimize icon rendering in RecursiveNavigation

Devanshusharma2005 committed:
cleanup...

Devanshusharma2005 committed:
Merge branch 'master' into docs/fix/sidebar-dropdown-indicator

Devanshusharma2005 committed:
Merge branch 'master' into docs/fix/sidebar-dropdown-indicator

Devanshusharma2005 commented:
@kiwicopple would love your review too. hehe ^^

Devanshusharma2005 committed:
Merge branch 'master' into docs/fix/sidebar-dropdown-indicator

Devanshusharma2005 committed:
Merge branch 'master' into docs/fix/sidebar-dropdown-indicator

charislam reviewed (commented):
This has been a problem for a while now, thanks for addressing it! ❤️ Looks really good, just some small feedback on top of @stylessh's

Devanshusharma2005 committed:
feat: enhance navigation with recursive dropdown indicators and performance optimizations

- Add visual chevron indicators for expandable sidebar items
- Implement recursive navigation system using existing NavMenuSection types
- Auto-expand accordion sections containing current active page on load/refresh
- Extract LinkContainer as independent component to prevent recreation on renders
- Fix React 19 compatibility issues with ref handling using useCallback pattern
- Resolve duplicate key warnings with improved key generation strategy
- Optimize accordion transition timing (500ms duration with ease-in-out) to prevent element overlap
- Preserve existing icon logic with hasLightIcon support for theme variants
- Maintain backward compatibility with existing navigation data structures

Performance improvements:
- Memoized recursive components to prevent unnecessary re-renders
- Stable component references for better React reconciliation
- Proper TypeScript interfaces with explicit prop definitions

UX improvements:
- Smooth accordion animations with physics-based easing
- Clear visual indicators for expandable menu items
- Automatic expansion of sections containing current page
- Professional-grade transition animations

Technical details:
- Uses existing NavMenuSection recursive type structure
- Implements containsActivePath helper for active page detection
- Leverages Radix UI accordion with proper data-state attributes
- Maintains all existing functionality including dynamic menu injection

Devanshusharma2005 committed:
Merge branch 'master' into docs/fix/sidebar-dropdown-indicator

Devanshusharma2005 committed:
Cleanup...

Devanshusharma2005 committed:
feat: implement cookie-based persisted state for navigation accordion

- Replace localStorage with cookie-based persistence for SSR compatibility
- Add proper cookie helpers with SameSite=Lax and 30-day expiration
- Implement initialization state to prevent hydration mismatches
- Maintain auto-expand functionality for active page sections
- Preserve user's manual accordion state across page refreshes and sessions
- Add debounced cookie updates (300ms) for performance optimization
- Use proper cookie naming convention: 'supabase-docs-nav-state'

Benefits:
- SSR compatible: Works with server-side rendering
- Cross-session persistence: Maintains state across browser sessions
- Better security: Cookies are more secure than localStorage
- Performance: Debounced updates prevent excessive cookie writes
- User experience: Seamless navigation state preservation

Devanshusharma2005 committed:
refactor: implement individual item-based persistence following reference pattern

- Replace global state management with individual item persistence
- Use sessionStorage with 'nav-expansion-' prefix for each item
- Follow the exact pattern from reference: usePersistedExpansionState hook
- Maintain auto-expansion for active page sections
- Simplify state management by removing complex global state
- Each accordion item manages its own expansion state independently
- Preserve user's manual toggle state across page refreshes
- Use sessionStorage instead of cookies for better performance

Benefits:
- Cleaner architecture: Each item manages its own state
- Better performance: No global state updates
- Simpler logic: Direct item-to-storage mapping
- Reference pattern compliance: Follows established patterns
- Individual control: Each section can be toggled independently
- Session persistence: Maintains state during browser session

Devanshusharma2005 committed:
refactor: implement URL-driven navigation state following Supabase docs pattern

- Replace sessionStorage persistence with URL-driven expansion state
- Follow the established Supabase docs pattern: URL as single source of truth
- Remove individual item persistence hooks in favor of pathname-based logic
- Implement useUrlDrivenExpansion hook that determines open sections from current URL
- Use getSectionsContainingPath to find all sections that should be expanded
- Remove manual toggle functionality - sections open/close based on URL navigation
- Maintain smooth transitions and visual indicators for expandable sections
- Ensure consistency and reliability by using URL as the definitive state source

Benefits:
- Reliability: URL is always the single source of truth
- Consistency: Matches existing Supabase docs navigation behavior
- Simplicity: No complex state management or storage concerns
- Performance: No localStorage/sessionStorage operations
- SSR Compatible: Works perfectly with server-side rendering
- Predictable: Navigation state is always consistent with current page

Devanshusharma2005 committed:
persistant state.

Devanshusharma2005 committed:
Merge branch 'master' into docs/fix/sidebar-dropdown-indicator

Devanshusharma2005 committed:
file delete

Devanshusharma2005 committed:
cleanup..

Devanshusharma2005 committed:
file cleanup

Devanshusharma2005 committed:
cleanup...

Devanshusharma2005 committed:
cleanup..

stylessh committed:
Merge branch 'master' into docs/fix/sidebar-dropdown-indicator

Devanshusharma2005 committed:
formatting..

charislam commented:
Thanks @Devanshusharma2005! The more localized updates are much easier to review ❤️ 

Just one small outstanding problem:

In some cases, the positioning of the button shifts as the accordion opens and closes

https://github.com/user-attachments/assets/fa072377-2f28-4a98-9f76-e7cfc5566b62

Devanshusharma2005 commented:
> Thanks @Devanshusharma2005! The more localized updates are much easier to review ❤️
> 
> Just one small outstanding problem:
> 
> In some cases, the positioning of the button shifts as the accordion opens and closes
> 
>  CleanShot.2025-09-26.at.16.11.17.mp4

@charislam This is just following the default behavior we have everywhere else. It's working fine for other nested items, so I figured keeping it consistent made sense. If I lock the chevron in a fixed position, it might actually be more annoying for users since they'd expect it to move with the content. Plus, as you can see, the border line is also stretching according to the collapse which was the original behaviour, so everything moves together as intended. But hey, I'll be happy to be wrong here! What do you think?

charislam commented:
@Devanshusharma2005 yep, I saw it was the old behavior that the width of the entire thing changes 😅 I think it was fine before because it was mostly the line shrinking/expanding which is relatively subtle and doesn't disturb the user experience, but the dropdown indicator shifting is just really odd

it's probably a simple CSS thing to make the inner element take up the full space, just need to find out where to apply it 😄

Devanshusharma2005 committed:
aww shit that it.

Devanshusharma2005 commented:
@charislam figured it out. It was a very tiny fix. Idk why it took me so much time to debug it. Maybe I need to re-visit my css skills haha ^^

Devanshusharma2005 committed:
Merge branch 'master' into docs/fix/sidebar-dropdown-indicator

Devanshusharma2005 committed:
Merge branch 'master' into docs/fix/sidebar-dropdown-indicator

Devanshusharma2005 committed:
Merge branch 'master' into docs/fix/sidebar-dropdown-indicator

charislam reviewed (approved):
Nice, this is great, thank you!
